File: frontend/src/app/core/services/search.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiResponse } from './auth.service';
import { UserResponse } from './user.service';
import { PostResponse, PagedResponse } from './post.service';

export interface TrendingHashtag {
    id: number;
    tag: string;
    usageCount: number;
    lastUsed: string;
}

export interface SearchResults {
    users?: PagedResponse<UserResponse>;
    posts?: PagedResponse<PostResponse>;
    [key: string]: any;
}

@Injectable({
    providedIn: 'root'
})
export class SearchService {
    private apiUrl = '/api/search';

    constructor(private http: HttpClient) { }

    searchAll(query: string, limit: number = 5): Observable<ApiResponse<SearchResults>> {
        const params = new HttpParams()
            .set('query', query)
            .set('limit', limit.toString());
        return this.http.get<ApiResponse<SearchResults>>(`${this.apiUrl}/all`, { params });
    }

    searchUsers(query: string, page: number = 0, size: number = 10): Observable<ApiResponse<PagedResponse<UserResponse>>> {
        const params = new HttpParams()
            .set('query', query)
            .set('page', page.toString())
            .set('size', size.toString());
        return this.http.get<ApiResponse<PagedResponse<UserResponse>>>(`${this.apiUrl}/users`, { params });
    }

    searchPosts(query: string, page: number = 0, size: number = 10): Observable<ApiResponse<PagedResponse<PostResponse>>> {
        const params = new HttpParams()
            .set('query', query)
            .set('page', page.toString())
            .set('size', size.toString());
        return this.http.get<ApiResponse<PagedResponse<PostResponse>>>(`${this.apiUrl}/posts`, { params });
    }

    getTrendingSearches(): Observable<ApiResponse<string[]>> {
        return this.http.get<ApiResponse<string[]>>(`${this.apiUrl}/trending`);
    }

    getTrendingHashtags(limit: number = 10): Observable<ApiResponse<TrendingHashtag[]>> {
        return this.http.get<ApiResponse<TrendingHashtag[]>>(`/api/hashtags/trending?limit=${limit}`);
    }

    advancedPostSearch(params: any): Observable<ApiResponse<PagedResponse<PostResponse>>> {
        let httpParams = new HttpParams();
        Object.keys(params).forEach(key => {
            if (params[key] !== null && params[key] !== undefined) {
                httpParams = httpParams.set(key, params[key].toString());
            }
        });
        return this.http.get<ApiResponse<PagedResponse<PostResponse>>>(`${this.apiUrl}/posts/advanced`, { params: httpParams });
    }

    advancedUserSearch(params: any): Observable<ApiResponse<PagedResponse<UserResponse>>> {
        let httpParams = new HttpParams();
        Object.keys(params).forEach(key => {
            if (params[key] !== null && params[key] !== undefined) {
                httpParams = httpParams.set(key, params[key].toString());
            }
        });
        return this.http.get<ApiResponse<PagedResponse<UserResponse>>>(`${this.apiUrl}/users/advanced`, { params: httpParams });
    }

    getRecentSearches(): Observable<ApiResponse<string[]>> {
        return this.http.get<ApiResponse<string[]>>(`${this.apiUrl}/recent`);
    }

    clearRecentSearches(): Observable<ApiResponse<void>> {
        return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/recent`);
    }

    removeRecentSearch(query: string): Observable<ApiResponse<void>> {
        return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/recent/${encodeURIComponent(query)}`);
    }

    getSearchSuggestions(query: string): Observable<ApiResponse<string[]>> {
        return this.http.get<ApiResponse<string[]>>(`${this.apiUrl}/suggestions?query=${encodeURIComponent(query)}`);
    }
}

File: backend/src/main/java/org/revature/revconnect/service/SearchService.java
package org.revature.revconnect.service;

import org.revature.revconnect.dto.response.PagedResponse;
import org.revature.revconnect.dto.response.PostResponse;
import org.revature.revconnect.dto.response.UserResponse;
import org.revature.revconnect.enums.PostType;
import org.revature.revconnect.enums.UserType;
import org.revature.revconnect.exception.BadRequestException;
import org.revature.revconnect.mapper.PostMapper;
import org.revature.revconnect.mapper.UserMapper;
import org.revature.revconnect.model.Hashtag;
import org.revature.revconnect.model.Post;
import org.revature.revconnect.model.User;
import org.revature.revconnect.repository.PostRepository;
import org.revature.revconnect.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Service
@RequiredArgsConstructor
@Slf4j
public class SearchService {

    private final UserRepository userRepository;
    private final UserMapper userMapper;
    private final PostRepository postRepository;
    private final PostMapper postMapper;
    private final HashtagService hashtagService;
    private final AuthService authService;

    private static final int MAX_RECENT = 15;
    private static final Map<Long, Deque<String>> RECENT_SEARCHES = new ConcurrentHashMap<>();

    @Transactional(readOnly = true)
    public Map<String, Object> searchAll(String query, int limit) {
        int safeLimit = Math.max(1, Math.min(limit, 50));
        log.info("Global search for: {}", query);
        saveRecentQuery(query);

        Page<User> usersPage = userRepository.searchPublicUsers(query, PageRequest.of(0, safeLimit));
        Page<Post> postsPage = postRepository.searchPosts(
                normalize(query), null, null, null, null, null, PageRequest.of(0, safeLimit));
        List<Hashtag> hashtags = hashtagService.search(query);

        List<UserResponse> users = usersPage.getContent().stream()
                .map(userMapper::toPublicResponse)
                .toList();
        List<PostResponse> posts = postsPage.getContent().stream()
                .map(postMapper::toResponse)
                .toList();
        List<String> tags = hashtags.stream()
                .map(Hashtag::getName)
                .toList();

        Map<String, Object> result = new HashMap<>();
        result.put("users", users);
        result.put("posts", posts);
        result.put("hashtags", tags);
        return result;
    }

    @Transactional(readOnly = true)
    public PagedResponse<UserResponse> searchUsers(String query, int page, int size) {
        saveRecentQuery(query);
        Page<User> users = userRepository.searchPublicUsers(query, PageRequest.of(page, size));
        return PagedResponse.fromEntityPage(users, userMapper::toPublicResponse);
    }

    @Transactional(readOnly = true)
    public PagedResponse<PostResponse> searchPosts(String query, int page, int size) {
        saveRecentQuery(query);
        Page<Post> posts = postRepository.searchPosts(
                normalize(query), null, null, null, null, null, PageRequest.of(page, size));
        return PagedResponse.fromEntityPage(posts, postMapper::toResponse);
    }

    @Transactional(readOnly = true)
    public PagedResponse<PostResponse> advancedPostSearch(String query, String author, String dateFrom, String dateTo,
                                                          String postType, Integer minLikes, int page, int size) {
        saveRecentQuery(query);
        PostType parsedPostType = parsePostType(postType);
        LocalDateTime from = parseDateStart(dateFrom);
        LocalDateTime to = parseDateEnd(dateTo);

        Page<Post> posts = postRepository.searchPosts(
                normalize(query),
                normalize(author),
                parsedPostType,
                minLikes,
                from,
                to,
                PageRequest.of(page, size));

        return PagedResponse.fromEntityPage(posts, postMapper::toResponse);
    }

    @Transactional(readOnly = true)
    public PagedResponse<UserResponse> advancedUserSearch(String query, String location, String userType,
                                                          Boolean verified, int page, int size) {
        saveRecentQuery(query);
        UserType parsedUserType = parseUserType(userType);
        Page<User> users = userRepository.advancedSearchPublicUsers(
                normalize(query),
                normalize(location),
                parsedUserType,
                verified,
                PageRequest.of(page, size));
        return PagedResponse.fromEntityPage(users, userMapper::toPublicResponse);
    }

    public List<String> getSearchSuggestions(String query) {
        saveRecentQuery(query);
        return hashtagService.search(query).stream()
                .map(Hashtag::getName)
                .limit(10)
                .toList();
    }

    public List<String> getTrendingSearches() {
        return hashtagService.getTrending(10).stream()
                .map(Hashtag::getName)
                .toList();
    }

    public List<String> getRecentSearches() {
        Long userId = authService.getCurrentUser().getId();
        Deque<String> deque = RECENT_SEARCHES.getOrDefault(userId, new LinkedList<>());
        return new ArrayList<>(deque);
    }

    public void clearRecentSearches() {
        Long userId = authService.getCurrentUser().getId();
        RECENT_SEARCHES.remove(userId);
    }

    public void removeRecentSearch(String query) {
        if (query == null || query.isBlank()) {
            return;
        }
        Long userId = authService.getCurrentUser().getId();
        Deque<String> deque = RECENT_SEARCHES.get(userId);
        if (deque == null) {
            return;
        }
        String normalized = query.trim();
        LinkedHashSet<String> unique = new LinkedHashSet<>(deque);
        unique.remove(normalized);
        RECENT_SEARCHES.put(userId, new LinkedList<>(unique));
    }

    private String normalize(String value) {
        return value == null || value.isBlank() ? null : value.trim();
    }

    private PostType parsePostType(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        try {
            return PostType.valueOf(value.trim().toUpperCase());
        } catch (IllegalArgumentException ex) {
            throw new BadRequestException("Invalid postType: " + value);
        }
    }

    private UserType parseUserType(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        try {
            return UserType.valueOf(value.trim().toUpperCase());
        } catch (IllegalArgumentException ex) {
            throw new BadRequestException("Invalid userType: " + value);
        }
    }

    private LocalDateTime parseDateStart(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        try {
            return LocalDate.parse(value.trim()).atStartOfDay();
        } catch (Exception ex) {
            throw new BadRequestException("Invalid dateFrom: " + value);
        }
    }

    private LocalDateTime parseDateEnd(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        try {
            return LocalDate.parse(value.trim()).plusDays(1).atStartOfDay().minusNanos(1);
        } catch (Exception ex) {
            throw new BadRequestException("Invalid dateTo: " + value);
        }
    }

    private void saveRecentQuery(String query) {
        if (query == null || query.isBlank()) {
            return;
        }
        Long userId = authService.getCurrentUser().getId();
        String normalized = query.trim();
        Deque<String> deque = RECENT_SEARCHES.computeIfAbsent(userId, ignored -> new LinkedList<>());
        deque.remove(normalized);
        deque.addFirst(normalized);
        while (deque.size() > MAX_RECENT) {
            deque.removeLast();
        }
    }
}
